From: Simon McVittie <smcv@debian.org>
Date: Sun, 15 Feb 2015 23:51:23 +0000
Subject: Add support for DP_LINK_VORBIS make variable

Like the existing DP_LINK_TO_JPEG, this links libvorbisfile in the
normal way instead of dlopening it. Linux distributions like Debian
strongly prefer normal library linking: we have a lot of infrastructure for
tracking correct versioned dependencies in this mode of linking and
avoiding incompatible versions, and package systems like apt mean that
it's trivial to install dependencies.

This uses the existing support for linking to libvorbisfile in the normal
way, which is used on Android and iOS.  The default on other platforms
is the current behaviour: dlopen libvorbisfile.

Origin: vendor, Debian
Forwarded: no
---
 BSDmakefile  | 10 ++++++++++
 makefile     | 19 +++++++++++++++++++
 makefile.inc |  6 +++---
 3 files changed, 32 insertions(+), 3 deletions(-)

diff --git a/BSDmakefile b/BSDmakefile
index d13013b..a206678 100644
--- a/BSDmakefile
+++ b/BSDmakefile
@@ -28,6 +28,7 @@ DP_LINK_JPEG?=shared
 DP_LINK_ODE?=dlopen
 DP_LINK_CRYPTO?=dlopen
 DP_LINK_CRYPTO_RIJNDAEL?=dlopen
+DP_LINK_VORBIS?=dlopen
 
 ###### Optional features #####
 DP_CDDA?=enabled
@@ -126,6 +127,15 @@ LIB_CRYPTO_RIJNDAEL=
 CFLAGS_CRYPTO_RIJNDAEL=
 .endif
 
+# vorbis
+.if $(DP_LINK_VORBIS) == "shared"
+CFLAGS_LIBVORBIS=-DLINK_TO_LIBVORBIS `pkg-config --cflags vorbisfile`
+LIB_VORBIS=`pkg-config --libs vorbisfile`
+.else
+CFLAGS_LIBVORBIS=
+LIB_VORBIS=
+.endif
+
 .endif
 
 
diff --git a/makefile b/makefile
index 59504d1..c1f7d23 100644
--- a/makefile
+++ b/makefile
@@ -126,6 +126,7 @@ ifeq ($(DP_MAKE_TARGET), linux)
 	DP_LINK_ODE?=dlopen
 	DP_LINK_CRYPTO?=dlopen
 	DP_LINK_CRYPTO_RIJNDAEL?=dlopen
+	DP_LINK_VORBIS?=dlopen
 endif
 
 # Mac OS X configuration
@@ -161,6 +162,7 @@ ifeq ($(DP_MAKE_TARGET), macosx)
 	DP_LINK_ODE?=dlopen
 	DP_LINK_CRYPTO?=dlopen
 	DP_LINK_CRYPTO_RIJNDAEL?=dlopen
+	DP_LINK_VORBIS?=dlopen
 
 	# on OS X, we don't build the CL by default because it uses deprecated
 	# and not-implemented-in-64bit Carbon
@@ -202,6 +204,7 @@ ifeq ($(DP_MAKE_TARGET), sunos)
 	DP_LINK_ODE?=dlopen
 	DP_LINK_CRYPTO?=dlopen
 	DP_LINK_CRYPTO_RIJNDAEL?=dlopen
+	DP_LINK_VORBIS?=dlopen
 endif
 
 # BSD configuration
@@ -237,6 +240,7 @@ endif
 	DP_LINK_ODE?=dlopen
 	DP_LINK_CRYPTO?=dlopen
 	DP_LINK_CRYPTO_RIJNDAEL?=dlopen
+	DP_LINK_VORBIS?=dlopen
 endif
 
 # Win32 configuration
@@ -296,6 +300,7 @@ ifeq ($(DP_MAKE_TARGET), mingw)
 	DP_LINK_ODE?=dlopen
 	DP_LINK_CRYPTO?=dlopen
 	DP_LINK_CRYPTO_RIJNDAEL?=dlopen
+	DP_LINK_VORBIS?=dlopen
 endif
 
 # set these to "" if you want to use dynamic loading instead
@@ -348,6 +353,20 @@ ifeq ($(DP_LINK_CRYPTO_RIJNDAEL), dlopen)
 	CFLAGS_CRYPTO_RIJNDAEL=
 endif
 
+ifndef PKG_CONFIG
+	PKG_CONFIG=pkg-config
+endif
+
+# vorbis
+ifeq ($(DP_LINK_VORBIS), shared)
+	CFLAGS_LIBVORBIS=-DLINK_TO_LIBVORBIS `$(PKG_CONFIG) --cflags vorbisfile`
+	LIB_VORBIS=`$(PKG_CONFIG) --libs vorbisfile`
+endif
+ifeq ($(DP_LINK_VORBIS), dlopen)
+	CFLAGS_LIBVORBIS=
+	LIB_VORBIS=
+endif
+
 ##### Sound configuration #####
 
 ifndef DP_SOUND_API
diff --git a/makefile.inc b/makefile.inc
index 9c66ed1..864e22a 100644
--- a/makefile.inc
+++ b/makefile.inc
@@ -171,7 +171,7 @@ OBJ_SDL= builddate.c sys_sdl.o vid_sdl.o thread_sdl.o $(OBJ_MENU) $(OBJ_SND_COMM
 
 
 # Compilation
-CFLAGS_COMMON=$(CFLAGS_MAKEDEP) $(CFLAGS_PRELOAD) $(CFLAGS_FS) $(CFLAGS_WARNINGS) $(CFLAGS_LIBZ) $(CFLAGS_LIBJPEG) $(CFLAGS_D3D) $(CFLAGS_NET) $(CFLAGS_SDL) -D_FILE_OFFSET_BITS=64 -D__KERNEL_STRICT_NAMES -I../../../
+CFLAGS_COMMON=$(CFLAGS_MAKEDEP) $(CFLAGS_PRELOAD) $(CFLAGS_FS) $(CFLAGS_WARNINGS) $(CFLAGS_LIBZ) $(CFLAGS_LIBJPEG) $(CFLAGS_LIBVORBIS) $(CFLAGS_D3D) $(CFLAGS_NET) $(CFLAGS_SDL) -D_FILE_OFFSET_BITS=64 -D__KERNEL_STRICT_NAMES -I../../../
 CFLAGS_CLIENT=-DCONFIG_MENU -DCONFIG_CD $(CFLAGS_VIDEO_CAPTURE)
 CFLAGS_SERVER=
 CFLAGS_DEBUG=-ggdb
@@ -210,13 +210,13 @@ LDFLAGS_RELEASE=$(OPTIM_RELEASE) -DSVNREVISION=`{ test -d .svn && svnversion; }
 OBJ_GLX= builddate.c sys_linux.o vid_glx.o thread_pthread.o keysym2ucs.o $(OBJ_MENU) $(OBJ_SOUND) $(OBJ_CD) $(OBJ_VIDEO_CAPTURE) $(OBJ_COMMON)
 
 LDFLAGS_UNIXCOMMON=-lm $(LIB_ODE) $(LIB_Z) $(LIB_JPEG) $(LIB_CRYPTO) $(LIB_CRYPTO_RIJNDAEL)
-LDFLAGS_UNIXCL=-L$(UNIX_X11LIBPATH) -lX11 -lXpm -lXext -lXxf86vm -pthread $(LIB_SOUND)
+LDFLAGS_UNIXCL=-L$(UNIX_X11LIBPATH) -lX11 -lXpm -lXext -lXxf86vm -pthread $(LIB_SOUND) $(LIB_VORBIS)
 LDFLAGS_UNIXCL_PRELOAD=-lz -ljpeg -lpng -logg -ltheora -lvorbis -lvorbisenc -lvorbisfile -lcurl
 LDFLAGS_UNIXSV_PRELOAD=-lz -ljpeg -lpng -lcurl
 LDFLAGS_UNIXSDL_PRELOAD=-lz -ljpeg -lpng -logg -ltheora -lvorbis -lvorbisenc -lvorbisfile -lcurl
 CFLAGS_UNIX_PRELOAD=-DPREFER_PRELOAD
 
-LDFLAGS_UNIXSDL=$(SDLCONFIG_LIBS)
+LDFLAGS_UNIXSDL=$(SDLCONFIG_LIBS) $(LIB_VORBIS)
 EXE_UNIXCL=darkplaces-glx
 EXE_UNIXSV=darkplaces-dedicated
 EXE_UNIXSDL=darkplaces-sdl
